subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,721 @@
+			   const struct usb_setup_packet *const setup)
+{
+	const uint32_t duration = HID_GET_IDLE_DURATION(setup->wValue);
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		ddata->idle_rate = duration;
+	}
+
+	LOG_DBG("Set Idle, Report ID %u Duration %u", id, duration);
+
+	if (ops->set_idle != NULL) {
+		ops->set_idle(dev, id, duration * 4UL);
+	} else {
+		errno = -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int handle_get_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup,
+			   struct net_buf *const buf)
+{
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	uint32_t duration;
+
+	if (setup->wLength != 1U) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id != 0U && ops->get_idle == NULL) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		duration = ddata->idle_rate;
+	} else {
+		duration = ops->get_idle(dev, id) / 4UL;
+	}
+
+	LOG_DBG("Get Idle, Report ID %u Durationi %u", id, duration);
+	net_buf_add_u8(buf, duration);
+
+	return 0;
+}
+
+static int handle_set_report(const struct device *dev,
+			     const struct usb_setup_packet *const setup,
+			     const struct net_buf *const buf)
+{
+	const uint8_t type = HID_GET_REPORT_TYPE(setup->wValue);
+	const uint8_t id = HID_GET_REPORT_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (ops->set_report == NULL) {
+		errno = -ENOTSUP;
+		LOG_ERR("Set Report not supported");
+		return 0;
+	}
+
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
+		LOG_DBG("Set Report, Input Report ID %u", id);
+		errno = ops->set_report(dev, type, id, buf->len, buf->data);
+		break;
+	case HID_REPORT_TYPE_OUTPUT:
+		LOG_DBG("Set Report, Output Report ID %u", id);
+		errno = ops->set_report(dev, type, id, buf->len, buf->data);
+		break;
+	case HID_REPORT_TYPE_FEATURE:
+		LOG_DBG("Set Report, Feature Report ID %u", id);
+		errno = ops->set_report(dev, type, id, buf->len, buf->data);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return 0;
+}
+
+static int handle_get_report(const struct device *dev,
+			     const struct usb_setup_packet *const setup,
+			     struct net_buf *const buf)
+{
+	const uint8_t type = HID_GET_REPORT_TYPE(setup->wValue);
+	const uint8_t id = HID_GET_REPORT_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
+		LOG_DBG("Get Report, Input Report ID %u", id);
+		errno = ops->get_report(dev, type, id, net_buf_tailroom(buf), buf->data);
+		break;
+	case HID_REPORT_TYPE_OUTPUT:
+		LOG_DBG("Get Report, Output Report ID %u", id);
```

Comment:
Output report type cannot be get, this should be an error case.

Reply:
Just like with input report being set, HID Specification leaves it up to developer whether it is possible to get output report or not.